server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/DefaultMirrorAccessController.java

```diff
@@ -0,0 +1,182 @@
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Streams;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.server.internal.api.sysadmin.MirrorAccessControlRequest;
+import com.linecorp.centraldogma.server.internal.storage.repository.CrudRepository;
+import com.linecorp.centraldogma.server.internal.storage.repository.HasRevision;
+import com.linecorp.centraldogma.server.metadata.UserAndTimestamp;
+import com.linecorp.centraldogma.server.mirror.MirrorAccessController;
+
+public class DefaultMirrorAccessController implements MirrorAccessController {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultMirrorAccessController.class);
+
+    private static final UuidGenerator idGenerator = new UuidGenerator();
+
+    @Nullable
+    private CrudRepository<MirrorAccessControl> repository;
+
+    public DefaultMirrorAccessController(CrudRepository<MirrorAccessControl> repository) {
+        this.repository = repository;
+    }
+
+    public DefaultMirrorAccessController() {}
+
+    public void setRepository(CrudRepository<MirrorAccessControl> repository) {
+        checkState(this.repository == null, "repository is already set.");
+        this.repository = repository;
+    }
+
+    private CrudRepository<MirrorAccessControl> repository() {
+        checkState(repository != null, "repository is not set.");
+        return repository;
+    }
+
+    public CompletableFuture<MirrorAccessControl> add(MirrorAccessControlRequest request, Author author) {
+        return repository().save(MirrorAccessControl.from(request, author), author)
+                           .thenApply(HasRevision::object);
+    }
+
+    public CompletableFuture<MirrorAccessControl> update(MirrorAccessControlRequest request, Author author) {
+        return repository().update(MirrorAccessControl.from(request, author), author)
+                           .thenApply(HasRevision::object);
+    }
+
+    public CompletableFuture<MirrorAccessControl> get(String id) {
+        return repository().find(id).thenApply(HasRevision::object);
+    }
+
+    public CompletableFuture<List<MirrorAccessControl>> list() {
+        return repository().findAll().thenApply(list -> list.stream()
+                                                            .map(HasRevision::object)
+                                                            .collect(toImmutableList()));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> allow(String targetPattern, String reason, int order) {
```

Comment:
Is there any reason that you encourage using this via `PluginContext.mirrorAccessController()` instead of the REST API?

Reply:
It is possible to use REST API in a Plugin but it can be a hassle.
The plugin needs to manage an access token for each phase.

I was thinking of implementing a plugin to expose a callback as a REST API.
When the callback is invoked, a target pattern will be added to a whitelist or a blacklist via these APIs.

Reply:
> It is possible to use REST API in a Plugin but it can be a hassle.

What I meant was exposing this API via `MirrorAccessControlService`. Is there any reason for this?